### PR TITLE
WIP: Smart add tasks (discussion)

### PR DESCRIPTION
**Pull request by chimericdream**
This is not complete, but I have started working on it. I wanted to get feedback on the syntax of the smart add. I am currently looking at [Remember the Milk's structure](http://www.rememberthemilk.com/help/?ctx=basics.smartadd.howdoiuse) and basically using that as a basis, but thought it would be good to get other opinions on what might be the simplest solution for the most people.

Here's what I am thinking so far:

`Task name --"Some longer note here" ^due !difficulty #tag #tag #"tag name" $value + - *mtwrfsu`

Breaking it down:
- `Task name`: Self explanatory. This is the main text for the task
- `--"Some longer note here"`: The `--` would signify that the following string is the note for the task
- `^due`: (to-dos only) This is the syntax used by RTM, and I think it works fairly well. This would be a date (or possibly text like "tomorrow", "today", or "next thursday").
- `!difficulty`: (all except rewards) This would be 1, 1.5, or 2 currently. A better solution would be to parse the strings "easy", "medium", and "hard" (or similar.. "e", "m", "h", etc).
- `#tag`: Fairly self explanatory. Tags with spaces should be in quotes.
- `$value`: (rewards only) The amount, in gold, that the reward should cost
- `+` or `-`: (habits only) Whether the habit has a positive and/or negative action
- `*mtwrfsu`: (dailies only) The days the task should repeat on. I used the single-letter notation used by my old university for the letters.

Ideally the items that are validated against a preset list of values (tags, repeating days, difficulty, etc) could pop open with a select box similar to the ones on RTM when their special character is entered into the box.

Thoughts? The actual parsing should be fairly quick to add, but the smart popups will be a little trickier, and I didn't want to build something that I thought was good but didn't have widespread appeal.

### CODE REVIEW

**Comment by crookedneighbor** [comment]
This is something I've wanted for a while. 

For reference, [here's a trello card where the feature has been discussed](https://trello.com/c/vDYCrh16/13-add-new-parse-text-do-this-tag-hard-monday). That's not to say that what has been written on the trello card is how it should be approached, I just wanted it here for easy access.

**Comment by crookedneighbor** [comment]
I'd be especially excited about this if the autocomplete features were implemented. There's an [autocomplete controller](https://github.com/HabitRPG/habitrpg/blob/a8cb0fd4e9785f257ca5edbc9c62dba7577d5ca2/website/public/js/controllers/groupsCtrl.js#L174-L227) for auto-completing user's names in the chat, maybe some of that can be refactored into a directive so multiple controllers can use the autocomplete scripts.

**Comment by chimericdream** [comment]
I agree. I looked at the card the other day, but the suggestions there seem to be somewhat at odds with one another in many ways, though there are a few similarities. One person mentioned Remember the Milk, which I used to use, and that reminded me that they already have a good implementation of this kind of thing.

I'll take a look at the autocomplete controller. That's probably going to come in _very_ handy.

**Comment by crookedneighbor** [comment]
@chimericdream Another instance where it'd be helpful to have autocomplete is with emojis when writing in markdown. IE, start typing `:ang` and a popup shows 
- "angel :angel:"
- "anguished :anguished:"
- "anger :anger:"
- "angry :angry:"

**Comment by chimericdream** [comment]
Good idea. I'll add that to my list. :smile:

**Comment by chimericdream** [comment]
@crookedneighbor What are your thoughts on the syntax? I didn't see comments from you in the Trello card. Personally, I like the RTM approach, but I'd like to have at least one more person who agrees before I start down that road. :smiling_imp:

**Comment by crookedneighbor** [comment]
Overall, I like the unique symbol approach, especially since it allows you to put them in whatever order you want, as long as the task text is first. 

Just a couple of concerns/thoughts:

### Repeating Dates `*`

Since asterisks are used for `*phrase that should be italic*` and `**phrase that should be bold**` in markdown, we'd have to be careful not to misrepresent a user's choice to use them. 

Example task input:

```
My *best task ever*
```

We would need to make sure that it would be interpreted correctly. We wouldn't want the repeater autocomplete to pop up in that instance.

I think ultimately, it'd make more sense  to go with `^ (or some other symbol)` for due dates and repeating days. In the function, we could have an `if(task.type =='todo') // interpret as due date` and `else if(task.type == 'daily') // interpret as repeating day`.

### Tags `#`

The tag autocompleter should be disabled if `#` is used at the beginning of an input, since that would denote a heading in markdown.

Example task inputs:

```
#Task with large heading
###A smaller heading
```

### Value symbol `$`

It'd be cool if we could set it to use multiple currency symbols. We have a lot of international users.

I also don't want to prevent users from using the currency symbol in their task names.

### Languages that type right to left

Do you think it would be difficult to support this if the user has their language set to Hebrew?

### Habits `+` or `-`

I don't want to prevent users from using the `+` or `-` keys in the task name. I, for instance, am very fond of :+1:, which can be written as `:+1:`.

### Difficulty `!`

We'd have to make sure it didn't interfere with `image alt attribute` for use in task names.

**Comment by chimericdream** [comment]
All good points. I had forgotten (because I don't really use it) that tasks allow Markdown in their processing. I think some of the items would be easy to work around (e.g. difficulty, which must be followed by one of three specific values). I am not sure about the value symbol. Since it's not necessarily "US Dollars" but rather "Gold Pieces", perhaps using something totally different like "@15", "15GP", "5SP", etc would be better. As for RTL languages, I don't know. My hope is that regular expressions would be order-independent, but I don't know that for certain.

**Comment by Alys** [comment]
I like all the suggestions here, but something to keep in mind for difficulty and days of the week is that people using other languages will not be typing easy (or e), or m for Monday, etc.

**Comment by chimericdream** [comment]
@Alys That's definitely something I'll try to take into account. I think I will get a working proof of concept using English terms, then figure out how to generalize it to other translations.

**Comment by chimericdream** [comment]
I have updated the branch with a first version of the smart-add functionality. I have done a few basic tests for all four categories (habits, dailies, to-dos, and rewards), but would welcome further testing. Here are the simple values I tested with:

**Habit**: `Task name --"Some longer note here" #morning !medium + -`
**Daily**: `Task name --"Some longer note here" #afternoon !hard ^mwf`
**To-Do**: `Task name --"Some longer note here" #evening !easy ^3/2/15`
**Reward**: `Task name --"Some longer note here" #afternoon 15GP`

My next step is to take a look at the auto-complete controller to see what can be pulled out into a reusable directive. I believe that will take care of a significant amount of the issues surrounding translations, but can't be totally sure until I dig into it a little more.

I welcome feedback on how it functions so far.

**Edit**: I know that at least a few of the things brought up by @crookedneighbor are still not addressed in this version yet, but I wanted to get it to others in the rough form for extra input.

**Comment by chimericdream** [comment]
Minor update to merge the current `develop` branch into the smart add. Now you can smartly add multiple items at a time. :smile:

File: website/public/js/controllers/tasksCtrl.js [hunk 1]
```diff
@@ -24,13 +24,100 @@ habitrpg.controller("TasksCtrl", ['$scope', '$rootScope', '$location', 'User','N
...
+          case 'medium':
+            newTask.priority = 1.5;
+            break;
+          case 'easy':
+          default:
+            newTask.priority = 1;
+            break;
+        }
+      }
+      if (tags.length === 0) {
```
Comment (crookedneighbor) [inline website/public/js/controllers/tasksCtrl.js]: Cool stuff, but it's broken for putting in just a task name. Chrome console displays this for this line when input is `task name`:

```
TypeError: Cannot read property 'length' of null
    at addTask (http://localhost:3000/js/controllers/tasksCtrl.js:106:15)
    at Scope.$scope.addTask (http://localhost:3000/js/controllers/tasksCtrl.js:132:9)
    at $parseFunctionCall (http://localhost:3000/bower_components/angular/angular.js:12332:18)
    at ngEventDirectives.(anonymous function).compile.element.on.callback (http://localhost:3000/bower_components/angular/angular.js:21389:17)
    at Scope.$get.Scope.$eval (http://localhost:3000/bower_components/angular/angular.js:14384:28)
    at Scope.$get.Scope.$apply (http://localhost:3000/bower_components/angular/angular.js:14483:23)
    at HTMLFormElement.<anonymous> (http://localhost:3000/bower_components/angular/angular.js:21394:23)
    at HTMLFormElement.n.event.dispatch (http://localhost:3000/bower_components/jquery/dist/jquery.min.js:3:6444)
    at HTMLFormElement.n.event.add.r.handle (http://localhost:3000/bower_components/jquery/dist/jquery.min.js:3:3219)
```

File: website/public/js/controllers/tasksCtrl.js [hunk 2]
```diff
@@ -24,13 +24,100 @@ habitrpg.controller("TasksCtrl", ['$scope', '$rootScope', '$location', 'User','N
...
+        return;
+      };
+      var regexes = {
+        notes: /(?:\s+--"([^"]+)")/,
+        due: /(?:\s+\^(\d?\d\/\d?\d\/\d?\d?\d\d))/,
+        diff: /(?:\s+\!(easy|medium|hard))/,
+        tag: /(?:\s+\#([^"\s]+|"[^"]+"))/g,
+        value: /(?:\s+(\d+(?:\.\d{1,2})?|0?\.\d{1,2})?GP)/,
+        up: /(?:\s+(\+))/,
+        down: /(?:\s+(\-))/,
```
Comment (crookedneighbor) [inline website/public/js/controllers/tasksCtrl.js]: If no `+` or `-` is present, it creates a habit with no buttons. The default should stay the way it is now, with both buttons displaying if no specific direction is given.

File: website/public/js/controllers/tasksCtrl.js (hunk 1 above)
Comment (chimericdream) [inline website/public/js/controllers/tasksCtrl.js]: Wait, we still want people to be able to add a task _without_ specifying all of the metadata every time? Dang!

Sarcasm aside, good catch. I didn't think to test the most basic of cases. :flushed:

File: website/public/js/controllers/tasksCtrl.js (hunk 2 above)
Comment (chimericdream) [inline website/public/js/controllers/tasksCtrl.js]: I agree. It's apparent I definitely need to test a lot more cases than the few I've done so far. I agree about keeping the current default.

**Comment by chimericdream** [comment]
One thing that I have been thinking about and is looking more desirable based on @crookedneighbor's feedback so far is the overall syntax of the smart task. In the [Trello card](https://trello.com/c/vDYCrh16/13-add-new-parse-text-do-this-tag-hard-monday) there is talk of using `::` to separate the task name from the metadata. I wonder if it might be cleaner to adopt that syntax in addition to the stuff above. That way, there would be much less risk (theoretically none) of accidentally detecting metadata when some characters are used in a task's name.

So the above items would become something like this:

**Habit:** `Task name :: --"Some longer note here" #morning !medium + -`
**Daily:** `Task name :: --"Some longer note here" #afternoon !hard ^mwf`
**To-Do:** `Task name :: --"Some longer note here" #evening !easy ^3/2/15`
**Reward:** `Task name :: --"Some longer note here" #afternoon 15GP`

Obviously this doesn't fix all of the other issues, but it could simplify some parts.

File: website/public/js/controllers/tasksCtrl.js (hunk 2 above)
Comment (chimericdream) [inline website/public/js/controllers/tasksCtrl.js]: The habit should now be working without `+` and `-`.

**Comment by chimericdream** [comment]
I spent some time looking into the auto-complete this evening. The chat boxes use the [Angular At Directive](https://github.com/scarletsky/Angular-At-Directive) plugin. There is a [closed issue](https://github.com/scarletsky/Angular-At-Directive/issues/4) on that repository regarding multiple flags, which this feature would require. It seems that matching different flags (in this case, `#`, `^`, and `!`) is possible, but the insertion is hard-coded to use the `@` symbol.

The plugin looks like it may be abandoned, as it has not been updated in nearly a year and a half. I looked at the code, and it shouldn't be too difficult to fork it and add the functionality for arbitrary flags, but that would require changing the dependencies of the HabitRPG repo to use the new version of the plugin.

Alternatively, we could either look for another means of providing auto-complete functionality or build something new that is specifically geared toward our need but general enough for the different sections of the site.

@crookedneighbor what do you think?

**Comment by chimericdream** [comment]
I have done some more testing with my latest fixes, and so far it seems to be working with the syntax as described a few posts up. I haven't yet tested with things like markdown and `:+1:` type items, but this is progress.

I am going to move forward and fork the [At Directive](https://github.com/scarletsky/Angular-At-Directive) plugin to make the auto-complete functionality more robust. The original plugin hasn't been updated since it was uploaded in September 2013, so I don't see much of a down side. In the worst case, my fork becomes the unmaintained/obsolete version in a couple years. :grinning:

**Comment by chimericdream** [comment]
@crookedneighbor I would really like to add a number of unit tests around this so we can test many scenarios more quickly, but as you mentioned in another PR the tests are kind of a mess. What is the plan for them moving forward? I saw your work on the API tests. I have worked with Karma tests for AngularJS apps, but haven't done much with Coffee script.

**Comment by chimericdream** [comment]
Also, I should note that I tried adding a new spec file for the tasksCtrl controller, but cannot get the test suite to run on the VM at all.

**Comment by crookedneighbor** [comment]
@chimericdream we have some tests for the controllers using karma, but recent changes to the repo structure broke them. I've corrected the paths in the `karma.conf.js` file, but it can't seem to find the habitrpg module now. 

Here's the error from running `grunt karma`: 

```
Running "karma:unit" (karma) task
INFO [karma]: Karma v0.10.10 server started at http://localhost:8080/
INFO [launcher]: Starting browser Firefox
INFO [Firefox 35.0.0 (Mac OS X 10.10)]: Connected on socket XuerUV7UcU-GIuOAdJM3
Firefox 35.0.0 (Mac OS X 10.10) ERROR
    Error: [$injector:nomod] Module 'habitrpg' is not available! You either misspelled the module name or forgot to load it. If registering a module ensure that you specify the dependencies as the second argument.
    http://errors.angularjs.org/1.3.9/$injector/nomod?p0=habitrpg
    at /Users/blade/habitrpg/habitrpg/website/public/bower_components/angular/angular.js:63
Firefox 35.0.0 (Mac OS X 10.10): Executed 0 of 0 ERROR (0.527 secs / 0 secs)
```

Obviously, something isn't getting loaded in correctly. 

I don't have a lot of experience with Karma (but I'm hoping to change that), so if you want to troubleshoot it in a separate PR, you'll be my hero forever.

**Comment by crookedneighbor** [comment]
Regarding the VM issues, I think the main reason people needed the VM on windows was because of needing to symlink the habitrpg-shared repo to the main repo. Since that's no longer needed, you may be able to just install node.js and mongodb on your windows machine and run it. 

You definitely won't be able to run the karma tests from the VM.

**Comment by chimericdream** [comment]
@crookedneighbor I might take a look at the VM in general then (including tests and such). One of the nice things for me is that I routinely switch between multiple machines on different OSes, so having a Vagrant VM dramatically simplifies the process of ensuring that specific versions of all the different dependencies are installed.

I have another project that runs Karma tests on a VM, so I'll see where the two differ and what I can do to shore up both the tests and the Vagrant box.

On the present topic, what do you think of my suggestion regarding moving ahead with a fork of the Angular-At-Directive plugin so that we can get auto-complete working more generically?

**Comment by chimericdream** [comment]
I have been looking for an auto-complete directive that supports multiple different flags, along the lines of what we would need to implement this. I think I may have found a replacement in [Ment.io](). It supports multiple auto-completes on a single input, each with a different flag. You can take a look at [the demo](http://jeff-collins.github.io/ment.io/#/examples) (specifically the text input one) to see an example that's right up our alley. The relevant directives are at [line 97 of the template](https://github.com/jeff-collins/ment.io/blob/master/ment.io/index.html#L97).

I am going to add ment.io as an additional dependency for the smart add feature, and I think it is worth looking into replacing any usages of the at directive with this over time.

**Comment by crookedneighbor** [comment]
@chimericdream I like that so much better than our current autocomplete implementation. Good find! Anyone else have thoughts?

**Comment by chimericdream** [comment]
@crookedneighbor Glad you like it. I think the safest way to do it is to make sure it works for the smart-add feature and leave the existing auto-complete in place for now. That way we don't have any major issues with possible regressions from moving from one to the other.

Another up side is the fact that Ment.io is much more active than the AtDirective. The latter hasn't been updated since it was uploaded to GitHub over a year and a half ago. The former has 150 commits and counting. Definitely good news for us.

**Comment by chimericdream** [comment]
Running into issues with the ment.io plugin. I have posted on an existing (closed and "fixed") issue on that repo, but have not been able to track down a solution after several hours of trying. I'll put this on hold for a few days and look at some of the other things I have on my list to tackle.

I still think the ment.io plugin will be the best option for our needs, but the issue (as far as I can tell right now) is related to items using both the `mentio` (or `mentioMenu`) directive and `ng-if`. There is some kind of race condition that causes the directive to fail and basically do nothing. I'm not 100% sure this is the exact issue I am facing, but after reading through [the issue](https://github.com/jeff-collins/ment.io/issues/27) I believe it is likely to be related. I am also using dynamic IDs (for the four task input boxes), so that could be causing interesting reactions.

**Comment by crookedneighbor** [comment]
@chimericdream Any update on this?

**Comment by chimericdream** [comment]
@crookedneighbor I just posted again on the issue on the mentio GitHub page. I'll merge the latest `develop` later this week and try to remember how close I was. I seem to recall that the bug in the mentio library is blocking much more forward progress, and I remember not finding anything else that was suited to our particular set of needs.

**Comment by crookedneighbor** [comment]
This is definitely still a desired feature, but since this pull request is very stale, I'm going to close it. If you'd like to bring it up to date so we can continue reviewing it, I would be thrilled to do so.

**Pull request closed without merging**